src/java/com/zimbra/oauth/handlers/impl/FacebookContactsImport.java

```diff
@@ -0,0 +1,569 @@
+                final JsonNode jsonResponse = getContactsRequest(url);
+                respContent = jsonResponse.toString();
+                if (jsonResponse.has("error") && jsonResponse.get("error").isContainerNode()) {
+                    ZimbraLog.extensions.debug(
+                            "Error found in JSON response object while fetching contacts. Response body: %s",
+                            respContent);
+                    throw ServiceException.FAILURE("An error was found in the api response. "
+                            + jsonResponse.get("error").asText(""), null);
+                }
+                // log this only at the most verbose level, because this contains
+                // privileged information
+                if (jsonResponse != null && jsonResponse.isContainerNode()) {
+                    if (jsonResponse.has("data")
+                            && jsonResponse.get("data").isContainerNode()) {
+                        final JsonNode contactsObject = jsonResponse.get("data");
+                        handleApiResponse(existingContacts, contactsObject);
+                        // check for next page
+                        if (jsonResponse.has("paging")
+                                && jsonResponse.get("paging").has("next")) {
+                            nextPageUrl = jsonResponse.get("paging").get("next").asText();
+                        }
+                    } else {
+                        ZimbraLog.extensions
+                        .info("Did not find 'data' element in json object");
+                    }
+                } else {
+                    ZimbraLog.extensions.debug("Did not find JSON response object.");
+                }
+            } while (nextPageUrl != null);
+        } catch (UnsupportedOperationException | IOException e) {
+            ZimbraLog.extensions.debug(String.format(
+                    "Data source test failed. Failed to fetch contacts from Facebook Contacts API"
+                    + "for testing. Response body: %s",
+                    respContent), e);
+            throw ServiceException.FAILURE(String.format(
+                "Data source test failed. Failed to fetch contacts from Facebook Contacts API"
+                    + "for testing. Response body: %s",
+                respContent), e);
+        }
+
+    }
+
+    /**
+     * Handles contacts from the Facebook api excluding existing contacts in the datasource.
+     *
+     * @param existingContacts Existing contacts
+     * @param contactsObject JSON from Facebook api response
+     * @throws ServiceException If an error is encountered
+     */
+    protected void handleApiResponse(Set<String> existingContacts, JsonNode contactsObject)
+            throws ServiceException {
+
+        List<ParsedContact> contactList = new ArrayList<ParsedContact>();
+        parseNewContacts(existingContacts, contactsObject, "id", contactList);
```

Comment:
It seems like this method `handleApiResponse` should be part of `parseNewContacts` ?

Do the create in `importData` like the other implementations.